### ISSUE TITLE
`partial` plugin does not handle overloads

### ISSUE THREAD

**Issue by asottile-sentry** [python/mypy#17585]
**Bug Report**

it appears the new `partial` plugin added in mypy 1.11 doesn't properly handle overloads (and just picks the first one)

**To Reproduce**

```python
import functools
from typing import assert_type, overload

@overload
def identity(x: int, *, p: str) -> int: ...
@overload
def identity(x: str, *, p: str) -> str: ...
def identity(x: int | str, *, p: str) -> int | str: return x


p = functools.partial(identity, p='hi')
assert_type(p(1), int)
assert_type(p('hi'), str)
```

**Expected Behavior**

(no errors)

**Actual Behavior**

```console
$ mypy t.py 
t.py:13: error: Expression is of type "int", not "str"  [assert-type]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.11.0
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12.2

**Comment by asottile-sentry** [comment]
an alternative spelling with `TypeVar`s seems to not work either:

```python
import functools
from typing import assert_type, overload, TypeVar

T = TypeVar("T", int, str)

def identity(x: T, *, p: str) -> T: return x


p = functools.partial(identity, p='hi')
assert_type(p(1), int)
assert_type(p('hi'), str)
```

```console
$ ./venv/bin/mypy t.py 
t.py:10: error: Expression is of type "Any", not "int"  [assert-type]
t.py:11: error: Expression is of type "Any", not "str"  [assert-type]
Found 2 errors in 1 file (checked 1 source file)
```

**Comment by hauntsaninja** [comment]
Yes, the behaviour with overloads is unchanged from mypy 1.10. Shouldn't be too bad to add support for it.
https://github.com/python/mypy/blob/d1d3c780c7a2d30b2a038903289ea7487303a218/mypy/plugins/functools.py#L131-L132

The constrained type var example you have is a duplicate of https://github.com/python/mypy/issues/17411. Note that unconstrained type vars or bounded type vars should work correctly. In general, mypy's handling of constrained type vars is a little weird.

**Comment by AlexWaygood** [comment]
> Note that unconstrained type vars or bounded type vars should work correctly.

Hmm, that doesn't seem to be the case as far as I can see @hauntsaninja :( We're hitting this in our attempt to upgrade mypy to 1.11.1 for Argument Clinic over at CPython in https://github.com/python/cpython/pull/122550. Minimal repro:

```py
from typing import Callable, TypeVar
import functools

T = TypeVar("T", bound=Callable[[str, int], str])
S = TypeVar("S", bound=type[str])

def foo(f: T) -> T:
    g = functools.partial(f, "foo")  # error: "T" not callable  [misc]
    return f

def bar(f: S) -> S:
    g = functools.partial(f, "foo")  # error: "S" not callable  [misc]
    return f
```

https://mypy-play.net/?mypy=master&python=3.12&gist=93876fe629e5320a88349da7a9f65188

**Comment by AaronDMarasco** [comment]
I think I'm hitting this as well in mypy 1.18.2. I won't post the actual code, but the gist is that the overload should be chosen based on an enum choice:

- Have an `enum` "`CommandsEnum`" that describes a set of functions a function `commands_core` can perform
- Have `@overload`s of that function with the `command` set to something like `Literal[CommandsEnum.Command3]`
- Create a partial with `do_comm3 = partial(commands_core, command=CommandsEnum.Command3)`

It seems to always pick the "first" overload.
